<!DOCTYPE html>
<html>
<head>
   <meta charset="utf-8">
   <title>Filetypes - SciLifeLab Courses</title>

   <link rel="stylesheet" type="text/css" href="http://scilifelab.github.io/courses/css/style.css" >
   <link rel="stylesheet" type="text/css" href="http://scilifelab.github.io/courses/css/syntax.css" >

   <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
   <script src="http://scilifelab.github.io/courses/js/bootstrap.min.js"></script>
</head>
<body>
  <nav class="navbar navbar-inverse navbar-fixed-top">
      <div class="container">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="https://www.scilifelab.se/"><img src="http://scilifelab.github.io/courses/images/scilife-green_on_black-35px.png"></a>
        </div>
        <div id="navbar" class="collapse navbar-collapse">
          <ul class="nav navbar-nav">
            <li><a href="./">Home</a></li>
            <li><a href="http://scilifelab.github.io/courses/">Past Courses</a></li>
            <li><a href="http://www.scilifelab.se/education/courses/">Future Courses</a></li>
          </ul>
        </div><!--/.nav-collapse -->
      </div>
    </nav>

    <div class="container">
      <div class="content">
        <h1>File Types</h1>

<p><strong>NOTE:</strong> in syntax examples, the dollar sign ($) is not to be printed. The dollar sign is usually an indicator that the text following it should be typed in a terminal window.</p>

<h2>1. Connecting to UPPMAX</h2>

<p>The first step of this lab is to open a ssh connection to UPPMAX. You will need a ssh program to do this:</p>

<p>On Linux: it is included by default, named <strong>Terminal</strong>.</p>

<p>On OSX: it is included by default, named <strong>Terminal</strong>.</p>

<p>On Windows: <a href="http://bit.ly/19yaQOM">Google MobaXterm</a> and download it.</p>

<p>Fire up the available ssh program and enter the following (replace <strong>username</strong> with your uppmax user name). -X means that X-forwarding is activated on the connection, which means graphical data can be transmitted if a program requests it, i.e. programs can use a graphical user interface (GUI) if they want to.</p>

<p><code>bash
$ ssh -X username@milou.uppmax.uu.se
</code></p>

<p>and give your password when prompted. As you type, nothing will show on screen. No stars, no dots. It is supposed to be that way. Just type the password and press enter, it will be fine.</p>

<p>Now your screen should look something like this:</p>

<p><img src="files/linux-intro/just-logged-in.jpg" alt=""></p>

<h2>2. Getting a node of your own</h2>

<p>Usually you would do most of the work in this lab directly on one of the login nodes at uppmax, but we have arranged for you to have one core each to avoid disturbances. This was covered briefly in the lecture notes.</p>

<p><code>bash
$ salloc -A g2015005 -t 07:00:00 -p core -n 1 --no-shell --reservation=g2015005_mon &amp;
</code></p>

<p>check which node you got (replace <strong>username</strong> with your uppmax user name)</p>

<p><code>bash
$ squeue -u username
</code></p>

<p>should look something like this</p>

<p><img src="files/linux-intro/allocation.png" alt=""></p>

<p>where <strong>q34</strong> is the name of the node I got (yours will probably be different). Note the numbers in the Time column. They show for how long the job has been running. When it reaches the time limit you requested (3 hours in this case) the session will shut down, and you will lose all unsaved data. Connect to this node from within uppmax.</p>

<p><code>bash
$ ssh -X q34 
</code></p>

<p><strong>Note:</strong> there is a uppmax specific tool called jobinfo that supplies the same kind of information as squeue that you can use as well (<code>$ jobinfo -u username</code>).</p>

<h2>3. Copying files needed for laboratory</h2>

<p>To be able to do parts of this lab, you will need some files. To avoid all the course participants editing the same file all at once, undoing each other&#39;s edits, each participant will get their own copy of the needed files. The files are located in the folder <strong>/proj/g2015005/labs/filetypes</strong></p>

<p>Next, copy the lab files from this folder. -r means recursively, which means all the files including sub-folders of the source folder. Without it, only files directly in the source folder would be copied, NOT sub-folders and files in sub-folders.</p>

<p><strong>NOTE: Remember to tab-complete to avoid typos and too much writing.</strong></p>

<p>Ex.</p>

<p>```bash
$ cp -r <source> <destination></p>

<p>$ cp -r /proj/g2015005/labs/filetypes ~/glob/ngs-intro/
```</p>

<p>Have a look in <strong>~/glob/ngs-intro/uppmax_tutorial:</strong></p>

<p><code>bash
$ cd ~/glob/ngs-intro/filetypes
$ tree
</code></p>

<p>This will print a file tree, which gives you a nice overview of the folders where you are standing. As you can see, you have a couple of files and a couple of empty folders. In the <strong>0_ref</strong> folder you have a reference genome in fasta format and annotations for the genome in GTF format. In <strong>0_seq</strong> you have a fastq file containing the reads we will align.</p>

<h2>4. Running a mini pipeline</h2>

<p>The best way to see all the different file formats is to run a small pipeline and see which files we encounter along the way. The pipeline is roughly the same steps you&#39;ll do in the resequencing part of the course, so the focus now is not to learn how the programs work. The data is from a sequencing of the adenovirus genome, which is tiny compared to the human genom (36kb vs 3gb).</p>

<p>The starting point of the pipeline is reads fresh from the sequencing machine in fastq format, and a reference genome in fasta format. The goal of the exercise is to look at our aligned reads in a genome viewer together with the annotations of the adenovirus genome.</p>

<p>First, let&#39;s go through the steps of the pipeline:</p>

<ol>
<li><strong>Build an index for the reference genome.</strong>

<ul>
<li>This will speed up the alignment process. Not possible to do without it.</li>
</ul></li>
<li><strong>Align the reads.</strong>

<ul>
<li>Yepp.</li>
</ul></li>
<li><strong>Convert the SAM file to a BAM file.</strong>

<ul>
<li>We want to use the space efficiently.</li>
</ul></li>
<li><strong>Sort the BAM file.</strong>

<ul>
<li>We have to sort it to be able to index it.</li>
</ul></li>
<li><strong>Index the BAM file.</strong>

<ul>
<li>We have to index it to make it fast to access the data in the file.</li>
</ul></li>
<li><strong>View the aligned data together with the annotations.</strong></li>
</ol>

<p>Before we do any steps, we have to load the modules for the programs we will be running.</p>

<p><code>bash
$ module load bioinfo-tools bwa/0.7.8 samtools/1.1 IGV/2.3.17
</code></p>

<h3>1. Building an index</h3>

<ol>
<li><strong>Build an index for the reference genome.</strong></li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>

<p>All aligners will have to index the reference genome you are aligning your data against. This is only done once per reference genome, and then you reuse that index whenever you need it. All aligners have their own kind of index unfortunately, so you will have to build one index for each aligner you want to use. In this lab we will use BWA, so we will build a BWA index.</p>

<p>First, have a look in the 0_ref folder</p>

<p><code>bash
$ ll 0_ref
</code></p>

<p>You should see 2 files: the fasta file, the gtf file. Have a look at each of them with less, just to see how they look inside.</p>

<p>```bash
Syntax: bwa index <name of the fasta file you want to index></p>

<p>$ bwa index 0_ref/ad2.fa
```</p>

<p>Since the genome is so small this should only take a second or so. The human genome will probably take a couple of hours.</p>

<p>Look in the 0_ref folder again and see if anything has changed.</p>

<p><code>bash
$ ll 0_ref
</code></p>

<p>The new files you see are the index files created by BWA. We are now ready to align our reads.</p>

<h3>2. Align the reads</h3>

<ol>
<li>Build an index for the reference genome.</li>
<li><strong>Align the reads.</strong></li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>

<p>Now we have a reference genome that has been indexed, and reads that we should align. Do that using BWA, and give the arguments</p>

<p>```bash
Syntax: bwa aln <reference genome> <fastq file with reads> &gt; <name of the output file></p>

<p>$ bwa aln 0<em>ref/ad2.fa 0</em>seq/ad2.fq &gt; 1_alignment/ad2.sai
```</p>

<p>For some reason BWA doesn&#39;t create a SAM file directly. It creates a .sai file which contains only the coordinates for where each read aligns. This format can&#39;t be used by any other program, so we have to convert it to a SAM file. For this, we have to give the program the reference genome we aligned to, the reads we aligned, and the .sai file. It will use the data stored in these files to construct a SAM file.</p>

<p>```bash
Syntax: bwa samse <reference genome> <the .sai file> <the fastq reads file> &gt; <the sam file></p>

<p>$ bwa samse 0<em>ref/ad2.fa 1</em>alignment/ad2.sai 0<em>seq/ad2.fq &gt; 1</em>alignment/ad2.sam
```</p>

<p>This will create a SAM file in <strong>1_alignment</strong> called <strong>ad2.sam</strong>. Have a look at it with less.</p>

<h3>3. Convert to BAM</h3>

<ol>
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li><strong>Convert the SAM file to a BAM file.</strong></li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li>View the aligned data together with the annotations.</li>
</ol>

<p>The next step is to conver the SAM file to a BAM file. This is more or less just compressing the file, like creating a zip file. To do that we will use <strong>samtools</strong>, telling it that the input is in SAM format (<strong>-S</strong>), and that it should output in BAM format (<strong>-b</strong>).</p>

<p>```bash
Syntax: samtools view -S -b <sam file> &gt; <bam file></p>

<p>$ samtools view -S -b 1<em>alignment/ad2.sam &gt; 2</em>bam/ad2.bam 
```</p>

<p>Have a look in the <strong>2_bam</strong> folder.</p>

<p><code>bash
$ ll 2_bam
</code></p>

<p>The created BAM file is an exact copy of the SAM file, but stored in a much more efficient format.</p>

<h3>4. Sort and index the BAM file</h3>

<ol>
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li><strong>Sort the BAM file.</strong></li>
<li><strong>Index the BAM file.</strong></li>
<li>View the aligned data together with the annotations.</li>
</ol>

<p>A BAM file is taking up much less space than the SAM file, but we can still improve performace. An indexed BAM file is infinitely faster for programs to work with, but before we can index it, we have to sort it since it&#39;s impossible (no gains performace wise) to index an unsorted file.</p>

<p>To sort the BAMf file use the following command:</p>

<p>```bash
Syntax: samtools sort <bam file to sort> <prefix of the outfile></p>

<p>$ samtools sort 2<em>bam/ad2.bam 3</em>sorted/ad2.sorted
```</p>

<p>This will sort the ad2.bam file and create a new BAM file which is sorted, called <strong>ad2.sorted.bam</strong>. The .bam file ending will be added automatically, so you should not specify it, unless you want your file named ad2.sorted.bam.bam (you don&#39;t).</p>

<p>Now when we have a sorted BAM file, we can index it. Use the command</p>

<p>```bash
Syntax: samtools index <bam file to index></p>

<p>$ samtools index 3_sorted/ad2.sorted.bam
```</p>

<p>This will create an index named <strong>ad2.sorted.bam.bai</strong> . It&#39;s nicer to have the .bam and .bai named to samt &quot;prefix&quot;, so rename the .bai file to not have the .bam in its name.</p>

<p><code>bash
$ mv 3_sorted/ad2.sorted.bam.bai 3_sorted/ad2.sorted.bai
</code></p>

<h3>5. View the data with a genome viewer</h3>

<ol>
<li>Build an index for the reference genome.</li>
<li>Align the reads.</li>
<li>Convert the SAM file to a BAM file.</li>
<li>Sort the BAM file.</li>
<li>Index the BAM file.</li>
<li><strong>View the aligned data together with the annotations.</strong></li>
</ol>

<p>Now that we have to data aligned and prepared for easy access, we will view it in a genome viewer together with the annotations for the genome. Have a look at the annotations file with less.</p>

<p><code>bash
$ less -S 0_ref/ad2.gtf
</code></p>

<p>The <strong>-S</strong> will tell less to not wrap the lines, and instead show one line per line. If the line is longer than the window, you can user the left and right arraow to scroll to the left and right. Many tabular files are extemely more readable when using the -S option. Try viewing the file without it and see the difference.</p>

<p>To view the file, we will use the program <strong>IGV</strong> (Integrated Genome Viewer). You have already loaded the module for this program at the start of the lab, so start it by typing the following command (now we&#39;ll find out if you used -X in all your ssh connections!):</p>

<p><code>bash
$ igv.sh
</code></p>

<p>This will start IGV. There are 3 files we have to load in IGV. The first is the reference genome. Press the menu button located at <strong>&quot;Genomes - Load Genome from File...&quot;</strong> and find your reference genome in <strong>0_ref/ad2.fa</strong>.</p>

<p>The second file you have to load is the reads. Press the menu button <strong>&quot;File - Load from File...&quot;</strong> and select your <strong>3_sorted/ad2.sorted.bam</strong>.</p>

<p>The last fie you have to load is the annotation data. Press <strong>&quot;File - Load from File...&quot;</strong> again and select you annotation file in <strong>0_ref/ad2.gtf</strong>.</p>

<p>This will show you the reference genome, how all the reads are aligned to it, and all the annotation data. Try zooming in on an area and have a look at the reads and annotations. The figures you see in the picture are all derived from the data in the files you have given it. The reference genome, a fasta file containing the DNA sequence of the reference genome, is visible if you zoom to the smallest level. All the reads are drawn from the data in the BAM file using the chromosome name, the starting position and the ending position of each read. The annotation in GTF format are all plotted using the data in the GTF file. The chromosome name, the starting position, the ending position, and the additional information in the rest of the fields (strand, name of the annotation, etc).</p>

<h2>5. Create a CRAM file</h2>

<p>The CRAM format is even more efficient than the BAM format. To create a CRAM file, use samtools. Tell samtools that you want CRAM output (<strong>-C</strong>) and specify which reference genome it should use to do the CRAM conversion (<strong>-T</strong>)</p>

<p>```bash
Syntax: samtools view -C -T <reference genome> <bam file to convert> &gt; <name of cram file></p>

<p>$ samtools view -C -T 0<em>ref/ad2.fa 3</em>sorted/ad2.sorted.bam &gt; 4_cram/ad2.cram
```</p>

<p>Compare the sizes of the convered BAM file and the newly created CRAM file:</p>

<p><code>bash
$ ll -h 3_sorted/ad2.sorted.bam 4_cram/ad2.cram
</code></p>

<p>This will list both the files, and print the file size in a human readable format (<strong>-h</strong>). The CRAM file is roughly 1/3 of the size of the BAM file. This is probably because all the reads in the simulated data has the same quality value (BBBBBBBBBB). Fewer types of quality values are easier to compress, hence the amazing compression ratio. Real data will have much more diverse quality scores, and the CRAM file would be pethaps 0.7-0.8 time the original BAM file.</p>

      </div>
    </div><!-- /.container -->

   
	<!-- Google Analytics -->
	<script>
		(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
		(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
		m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
		})(window,document,'script','//www.google-analytics.com/analytics.js','ga');
		
		ga('create', 'UA-23985298-11', 'auto');
		ga('send', 'pageview');
	</script>
	<!-- End Google Analytics -->



</body>
</html>
